pkg/envoy/cds/response.go

```diff
@@ -63,42 +54,40 @@ func NewResponse(meshCatalog catalog.MeshCataloger, proxy *envoy.Proxy, _ *xds_d
 		log.Error().Err(err).Msgf("Failed to get local cluster config for proxy %s", proxyServiceName)
 		return nil, err
 	}
-	clusterFactories[localCluster.Name] = localCluster
+	clusters = append(clusters, localCluster)
 
+	// Add an outbound passthrough cluster for egress
 	if cfg.IsEgressEnabled() {
-		// Add a pass-through cluster for egress
-		passthroughCluster := getOutboundPassthroughCluster()
-		clusterFactories[passthroughCluster.Name] = passthroughCluster
-	}
-
-	for _, cluster := range clusterFactories {
-		log.Debug().Msgf("Proxy service %s constructed ClusterConfiguration: %+v ", proxyServiceName, cluster)
-		marshalledClusters, err := ptypes.MarshalAny(cluster)
-		if err != nil {
-			log.Error().Err(err).Msgf("Failed to marshal cluster for proxy %s", proxy.GetCommonName())
-			return nil, err
-		}
-		resp.Resources = append(resp.Resources, marshalledClusters)
+		clusters = append(clusters, getOutboundPassthroughCluster())
 	}
 
+	// Add an inbound prometheus cluster (from Prometheus to localhost)
 	if cfg.IsPrometheusScrapingEnabled() {
-		prometheusCluster := getPrometheusCluster()
-		marshalledCluster, err := ptypes.MarshalAny(&prometheusCluster)
-		if err != nil {
-			log.Error().Err(err).Msgf("Error marshaling Prometheus cluster for proxy with CN=%s", proxy.GetCommonName())
-			return nil, err
-		}
-		resp.Resources = append(resp.Resources, marshalledCluster)
+		clusters = append(clusters, getPrometheusCluster())
 	}
 
+	// Add an outbound tracing cluster (from localhost to tracing sink)
 	if cfg.IsTracingEnabled() {
-		tracingCluster := getTracingCluster(cfg)
-		marshalledCluster, err := ptypes.MarshalAny(&tracingCluster)
+		clusters = append(clusters, getTracingCluster(cfg))
+	}
+
+	resp := &xds_discovery.DiscoveryResponse{
+		TypeUrl: string(envoy.TypeCDS),
+	}
+
+	alreadyAdded := mapset.NewSet()
```

Comment:
@shashankram here is the de-duplication mechanism (moved it so it is closer to the marshaling)